File: src/main/java/com/github/dkoval/leetcode/challenge/SubarraySumsDivisibleByK.java
package com.github.dkoval.leetcode.challenge;

import java.util.HashMap;
import java.util.Map;

/**
 * <a href="https://leetcode.com/problems/subarray-sums-divisible-by-k/">Subarray Sums Divisible by K</a>
 * <p>
 * Given an integer array nums and an integer k, return the number of non-empty subarrays that have a sum divisible by k.
 * <p>
 * A subarray is a contiguous part of an array.
 * <p>
 * Constraints:
 * <ul>
 *  <li>1 <= nums.length <= 3 * 10^4</li>
 *  <li>-104 <= nums[i] <= 10^4</li>
 *  <li>2 <= k <= 10^4</li>
 * </ul>
 */
public interface SubarraySumsDivisibleByK {

    int subarraysDivByK(int[] nums, int k);

    class SubarraySumsDivisibleByKRev2 implements SubarraySumsDivisibleByK {

        @Override
        public int subarraysDivByK(int[] nums, int k) {
            // sum(nums[0 : i]) % k -> count
            Map<Integer, Integer> seen = new HashMap<>();
            seen.put(0, 1);

            int total = 0;
            int sum = 0;
            for (int x : nums) {
                sum += x;
                sum %= k;

                // corner case: handle mod of negative numbers
                if (sum < 0) {
                    sum += k;
                }

                int count = 0;
                if (seen.containsKey(sum)) {
                    count = seen.get(sum);
                    total += count;
                }
                seen.put(sum, count + 1);
            }
            return total;
        }
    }

    class SubarraySumsDivisibleByKRev3 implements SubarraySumsDivisibleByK {

        @Override
        public int subarraysDivByK(int[] nums, int k) {
            // sum(nums[0 : i]) % k -> count
            int[] seen = new int[k];
            seen[0] = 1;

            int total = 0;
            int sum = 0;
            for (int x : nums) {
                sum += x;
                sum %= k;

                // corner case : handle mod of negative numbers
                if (sum < 0) {
                    sum += k;
                }

                if (seen[sum] > 0) {
                    total += seen[sum];
                }
                seen[sum]++;
            }
            return total;
        }
    }
}

File: src/main/kotlin/com/github/dkoval/leetcode/problems/SubarraySumsDivisibleByK.kt
package com.github.dkoval.leetcode.problems

import com.github.dkoval.leetcode.challenge.SubarraySumsDivisibleByK

object SubarraySumsDivisibleByKRev1 : SubarraySumsDivisibleByK {

    // Resource: https://www.youtube.com/watch?v=u9m-hnlcydk
    override fun subarraysDivByK(nums: IntArray, k: Int): Int {
        val counts = IntArray(k)
        var sum = 0
        for (x in nums) {
            // (sum + x) % K
            sum += x % k
            sum %= k
            if (sum < 0) {
                sum += k
            }
            counts[sum]++
        }

        var ans = counts[0]
        for (c in counts) {
            ans += c * (c - 1) / 2
        }
        return ans
    }
}